Fix getting-started experience for new custom generator

dazuma opened the pull request:
Going through the process of creating a "new" subgenerator and running it against a proto. Collecting fixes here.

blowmage commented on shared/output/gapic/gems/my_plugin/lib/google/gapic/generators/my_plugin_generator.rb:
```diff
@@ -38,22 +38,6 @@ def initialize api
-        #   Google::Protobuf::Compiler::CodeGeneratorResponse::File>]
-        #   The files that were generated for the API.
-        def generate
-          files = []
```
For cloud we have this method and call `super`.

dazuma replied:
Added it back as a stub.

dazuma commented:
Changes:

* Updated the README instructions to fix a few errors. For example:
    * It was missing a `bundle exec rake update` in the instructions on building and installing the gem
    * The example command line for building the vision API was missing a `--ruby_gapic_opt=configuration=/path/to/vision.yml`
    * The installation system envrionment variable in the protoc installation instructions was unused.
* Wrote a new section in the README describing how to get started with a custom generator.
* Noted in the README that protoc can be installed easily using homebrew.
* Fixed some issues in the `gapic-generator` binary. For example:
    * Templates were copied into a subdirectory named literally `gem_name` instead of the actual gem name
    * There was no git repo for a generated client, which caused bundle install to fail
* Changed the `gapic-generator` usage to split `--output` into a separate argument, because I found requiring an array for `-g` to be clumsy.
* `gapic-generator` no longer overwrites an existing directory. (Previously it would abort only if a _file_ was blocking the destination path.)
* Implemented validity check on the custom generator name.
* The generated `Gemfile` for a custom generator no longer includes a hardcoded path for the `gapic-generator` dependency, because we want the new gem to be able to live outside the `gapic-generator-ruby` repo.
* The generator subclass for a new custom generator inherits from `DefaultGenerator` instead of `BaseGenerator`.
* We no longer copy the default templates directly into a new custom generator's `templates/default` directory (where they are not used anyway because the custom generator looks inside `templates/#{gem_name}`). Instead, they are copied into `example_templates` to emphasize that they are fyi only and are not used directly.

blowmage commented on README.md:
```diff
@@ -8,18 +8,25 @@ but there are currently no guarantees of stability or support.
 ### Install the Proto Compiler
 This generator relies on the Protocol Buffer Compiler to [orchestrate] the
-client generation.
+client generation. Install protoc version 3.7 or later.
```
Do we still need to have `protoc` installed now that we are using `grpc-tools`?

dazuma replied:
I honestly have no idea. Where are we installing `grpc-tools`?

blowmage replied:
It is a development dependency on gapic-generator. All the Rake tasks use it instead of protoc because it is much easier to use than installing protoc on the CI machines. So we removed the steps to install protoc for development, but they are still there for actually using the generator.

dazuma replied:
Hmm. If the `grpc-tools` binaries are sufficient across platforms, it might be worth making it the preferred implementation for generator usage as well (i.e. make it a runtime dependency of `gapic-generator`). I think we can look into that separately, apart from this PR.
